source/Cosmos.HAL2/Drivers/PCI/Network/RTL8168.cs

```diff
@@ -0,0 +1,358 @@
+
+                mRxDescriptor.Write32(xOffset + 12, 0);
+                mTxDescriptor.Write32(xOffset + 12, 0);
+
+            }
+        }
+
+        public RTL8168(PCIDevice device) : base()
+        {
+            if (device == null)
+            {
+                throw new ArgumentException("PCI Device is null. Unable to get Realtek 8168 card");
+            }
+            pciCard = device;
+
+            // We are handling this device
+            pciCard.Claimed = true;
+
+            BaseAddress = pciCard.BaseAddressBar[0].BaseAddress;
+
+            // Enable the card
+            pciCard.EnableDevice();
+
+            SetIrqHandler(device.InterruptLine, HandleNetworkInterrupt);
+
+            Ports.OutB((ushort)(BaseAddress + 0xE0), 0x08);
+
+            Reset();
+
+            // Get the MAC Address
+            byte[] eeprom_mac = new byte[6];
+            for (uint b = 0; b < 6; b++)
+            {
+                eeprom_mac[b] = Ports.InB((ushort)(BaseAddress + b));
+            }
+
+            mac = new MACAddress(eeprom_mac);
+
+            InitBuffers();
+
+            Ports.OutD((ushort)(BaseAddress + 0x44), 0x0000E70F); // Enable RX
+
+            Ports.OutD((ushort)(BaseAddress + 0x37), 0x04);
+
+            Ports.OutD((ushort)(BaseAddress + 0x40), 0x03000700); // Enable TX
+
+            Ports.OutD((ushort)(BaseAddress + 0xDA), 2048); // Max rx packet size
+
+            Ports.OutB((ushort)(BaseAddress + 0xEC), 0x3F); // No early transmit
+
+            Ports.OutD((ushort)(BaseAddress + 0x20), mTxDescriptor.Offset);
+            Console.WriteLine("addresstx desc: " + mTxDescriptor.Offset);
+
+            Ports.OutD((ushort)(BaseAddress + 0xE4), mRxDescriptor.Offset);
+            Console.WriteLine("addressrx desc: " + mRxDescriptor.Offset);
+
+            if (((GetMacVersion() & 0x7cf00000) == 0x54100000) || ((GetMacVersion() & 0x7cf00000) == 0x54000000))
+            {
+                Console.WriteLine("8168H Detected!");
+
+                Ports.OutD((ushort)(BaseAddress + 0x40), Ports.InD((ushort)(BaseAddress + 0x40)) | (1 << 7)); // AUTO TX FIFO
+            }
+
+            Ports.OutW((ushort)(BaseAddress + 0x3C), 0xC3FF); //Activating all Interrupts
+
+            Ports.OutB((ushort)(BaseAddress + 0x37), 0x0C); // Enabling receive and transmit
+
+            //Console.WriteLine("Netcard version: 0x" + System.Utils.Conversion.DecToHex((int)GetMacVersion() & 0x7cf00000));
+            //Console.WriteLine("Netcard version: 0x" + System.Utils.Conversion.DecToHex((int)GetMacVersion() & 0x7c800000));
+
+        }
+
+        public bool Reset()
+        {
+            Ports.OutB((ushort)(BaseAddress + 0x37), 0x10); /* Send the Reset bit to the Command register */
+            while ((Ports.InB((ushort)(BaseAddress + 0x37)) & 0x10) != 0) { } /* Wait for the chip to finish resetting */
+            return true;
+        }
+
+        protected void HandleNetworkInterrupt(ref IRQContext aContext)
+        {
```

Comment:
Are all the Console.WriteLines in this method required? They seem like debug output and should be removable